merlin-server/sql/merlin/tables/span.sql

```diff
@@ -20,6 +21,11 @@ create table span (
     foreign key (dataset_id, parent_id)
     references span
     on update cascade
+    on delete cascade,
+  constraint span_has_activity
```

Comment:
The reason I don't think we want to couple the span and activity tables is that they have potentially different lifecycles. If I delete an activity from a plan, that doesn't necessarily mean I want to delete the span from the simulation results.

Reply:
Fair point, so remove the cascading then, yeah?

Reply:
Ooh. If we remove the activity a span references, is there a way to tell PG to convert it from the activity id to null?

Reply:
Answer: yes. `on delete set null`

Reply:
I do think that's do-able (something like `on delete set null`), however I'm still wondering about the wisdom of doing that.

We have multiple independent actors here - someone editing the plan, and the merlin worker(s) running simulations. If we force plan updates to also update simulation results in a transaction, I wonder if we'll run into any database locking issues 🤔 I need to read up a bit to say for sure.

Reply:
> Alternatively, we could use a join table for this

This _might_ provide the decoupling we need - we'd need to think through how and when the join table gets updated

Reply:
I'm pondering this PR generally also. A knee jerk reaction is that a join table might be better here.

Reply:
What is the strategy here for changes to a plan and have simulation updates at the same time? Sounds like this is your key problem to all this questions. Ideally you don't want to have plan changes while a simulation is occurring. From the users' point of view, that would be inconvenient to have to wait for a simulation to finish before make any changes. One of the way to deal with this is to not really delete a plan but rather mark it deleted so that the simulation data still have something to hang onto. Some form of versioning approach. Do we decide on how this be done? That may leads you to the right solution here.

Reply:
@tungxbui I think you're suggesting we move the immutable approach all the way up to the plan level and create a new object for each change. This would definitely introduce data storage concerns.

Reply:
I'm going to pose a question pointed in the alternative direction. Do we really need immutable data downstream? Do we need to maintain multiple simulation outputs for a plan? Do we need to retain multiple expansion outputs for a given plan?

Or is there a valid case for use saying "there is a plan, a single simulation output for it that is invalidated and cleared on plan edit, a single command output for it that is cleared on the above clearing" such that we cascade mutation downstream.

I genuinely don't have any idea which approach fits our use case better. I know I concern about data traceability from a spacecraft ops perspective, but that might be something we push down to the missions to maintain if they need it.

Reply:
@dyst5422, that is just one way but I think we have to come up with a reasonable way to deal with this and then handle consistently from there on. Otherwise we will be back to this problem again. The data can be clean up after when the simulation is done and it detects that the plan has now obsolete. That is another way to keep the data storage down. But otherwise we might end up with data integrity problem when you have multiple actors change various part of the plan at the same time. It will be hard to keep the data consistent as one does not know what the other does.

Reply:
Hmmm, I don't even aware that we retain multiple expansion outputs for a plan. That is a good question. This sounds like even more difficult to keep everything straight. I don't have answer to that. Do we have a requirement for this or just something we think it maybe a good idea?

Reply:
Right now, I think its a result of maintaining multiple simulation results, which directly drive the command generation

Reply:
@dyst5422, I do like this idea:  "we move the immutable approach all the way up to the plan level." I think it will keep thing consistent. Also, that will keep the history of the changes as you suggest that missions may want to have this. The clean up can be done as part of operation process. As of right now, you can't really tell which change comes first, which comes later or even what exactly is the change as part of it can be overwritten by later changes.

Reply:
> In my mind, we've kind of got this hybrid approach because of how we are representing a plan in the UI (and thus in great part the mental model we adopt) as both a `plan` and as a simulation result (displaying activities and spans equivalently). I'd be interested in how a discussion of the approach there as it relates to our data model might result in conscious decisions on the data mutability model.

I rather deeply agree with you here. One of the things that's bothered me for a while is that, in the process of generating simulation results, we assign a `directiveId` to every span, even those produced by child activities (which thus don't have an associated *input* directive). We have to be careful to assign fresh IDs to children, but reuse the input IDs for root spans. IIRC (@camargo?), this makes it easier for the UI to look up the directive associated with a span, whilst still identifying each span uniquely. I would much rather give spans and directive independent ID spaces, and explicitly join between them. (Much as in my `directive_span` suggestion elsewhere in this PR discussion.)

In my mind, directives are point-like, and spans are durative; directives are specifications of intent, and spans are reports of effect. Previous planning tools conflate the two, and have been successful for it; but I'm not convinced the core data model needs to follow suit, even if our visualizations do.

> **Tung:** One of the ways to deal with this is to not really delete a plan but rather mark it deleted so that the simulation data still have something to hang onto. Some form of versioning approach

> **Dylan:** I'm going to pose a question pointed in the alternative direction. Do we really need immutable data downstream? Do we need to maintain multiple simulation outputs for a plan? Do we need to retain multiple expansion outputs for a given plan?

Plan versioning has been on our radar since... forever, really. I'm a big fan of an approach based on immutability, but to date we've justified the current approach as "a simulation is a snapshot of the effects of some prior version of the plan". You can see that in the tables involved, we maintain `revision` columns; so if you see a `simulation` referencing a revision of a `plan` that has a newer revision number, you know that the simulation is out of date, and that the input plan has been, per se, "lost".

I would be quite interested in a *snapshot*-based approach, where plan HEAD is always mutable, but resimulation requires a snapshot to be taken (whether automatically or manually). Planners could always go back and view a saved revision of a plan, and even delete it (also deleting its cached simulation results). Allowing the HEAD revision to be mutable also avoids a lot of the data retention issues, since it effectively coalesces any changes to a plan in between simulations.

(I would *not* worry about branching history for plan revisions at all at this point. For now, if someone wants to edit a saved plan snapshot, they can clone a new plan from that revision and begin working from its HEAD.)

Reply:
One big reason to save multiple simulations is to do a differential analysis between different options. You could do that with two separate plans, but that seems somewhat unnatural from a planner's perspective. In particular, saving simulation results as a planner refines their plan makes it much easier to validate that your changes have, in fact, improved the risk and reward profile. I believe you got this pretty much for free with APGen, since simulation results are/can be just output to the filesystem normally. It would be weird to invalidate old runs just because you changed the inputs in such a context.

Reply:
@Twisol, you have many good points there. It does confirm that we do need an approach in some shape or form to allow different versions of this data set to be retained and reviewed. A more complete versioning, it sounds like.